feat: project onboarding

cimigree opened the pull request:
closes: #1291 
# Project Onboarding: screens, navigation, and tests

This PR introduces the **Project Onboarding** feature and a small set of **presentation components** so I can reuse existing screens.

### New presentation components

* **`StartNewProjectBaseScreen`**

  * Pure, stateless UI. Accepts `TopIcon`, `onGoBack`, and `onStart`.
  * Lets us reuse the same screen layout with different icons (composition over abstraction).
  * Keeps props limited and agnostic (no leaking `navigation` into the presentational layer).

* **`CreateProjectForm`**

  * Handles only the “project name” field + passes an action for the submit.
  * Displays “Keep in mind” info rows.
  * Exposes `(projectName: string) => void` submit handler and `isPending` to block double taps.
  * Keeps form concerns decoupled from navigation / mutation logic.

### New onboarding screens

* **`Join a Project`** 

  * Just has text and a Go Back button

* **`StartNewProject`** (container)

  * Renders `StartNewProjectBaseScreen`.

* **`CreateProject`** (container)

  * Renders `CreateProjectForm`.
  * Calls `useCreateProject`, sets active project on success, and then because there is now a project, the MapScreen and regular app stack populates automatically.
  * I have no information in the ticket about which screens I should add next, though there is UX and info from design for the future (I think?). So I was not sure what to do here.
  * Blocks back nav while the mutation is pending and surfaces errors to `ErrorBottomSheet`.
  
* **`MapOnOwn`**

  * Creates a default project (localized name) and sets it active.
  * Blocks repeat taps and back nav while pending.
  * Lets the **RootStack** naturally flip to the default app flow once a project exists.

### Navigation

* **Root stack branching**

  * If device not named → onboarding (existing).
  * If device named **and** no projects → **project onboarding** (new).
  * Else → default app.

* Moved the ErrorBottomSheet to the RootStack so it could be used in the ProjectOnboarding and in the regular App Screens

## Tests Added

### Integration test (Jest + RTL)

* adds multiple tests in the RootStack test for checking the functioning of creating a project and checking for various error states and edge cases

### E2E test (Appium/Webdriver/BrowserStack)

* **`tests/e2e/specs/onboarding/project-onboarding.test.ts`**
* This tests the different project onboarding screens back and forth but does not create a project
* Instead creates a default solo project so we can keep the rest of the flow
* Updated the documentation

cimigree committed:
Adds new screen about joining a project. Adds a new screen for starting a project. Navigates to those screens. Make svgs color agnostic.

cimigree committed:
Updates the start new project screens. Adds a presentation create project screen to share. Uses that shared component to create a project in project onboarding.

cimigree committed:
Adds Map on Own screen.

cimigree committed:
Adds integration and End to End test.

cimigree committed:
Updates E2E documentation.

cimigree committed:
Adds test id to error bottom sheet. Moves error bottom sheet up in the nav tree. Adds a whole bunch of error and edge case tests for the rootstack and project onboarding.

cimigree committed:
Merges with first branch

cimigree committed:
Adds suspense to regular app.

cimigree commented:
@ErikSin I had to wrap the Provider in the main app with suspense in order to get the one test where I create a project to pass. It was not an issue of adding Suspense to the test itself which makes me think it is actually needed in the code and would be useful for real world use cases. It also passed fine when I had suspense in the Context, but I know you did not want it there.

ErikSin commented on src/frontend/sharedComponents/Projects/CreateProjectForm.tsx:
I think we need to reconsider how we are reusing shared UI.

We are doing two different things, both which independently solve the same problem, but we are combining them which makes it a little confusing

Originally The `CreateOrJoinProject` rendered the same screen, and we conditionally changed the text and action based on the navigation name.

Now you have created this component, which does the same thing as above, which is to allow for the same UI with slight variations on the text and action.

But now the CreateOrJoinProject uses this component, so were using 2 different tactics to solve the same problem, when only one is needed.

I also think the 2 create project screens should actually be the same. I just sent a [message to design](https://awanadigital.slack.com/archives/C6ZMBFATZ/p1755197488069659) to ask. 

If the 2 create project screen are the same, I am still in favor of using the original tactic of one screen conditionally rendered based on the nav name. While the code is slightly more verbose, it avoids nesting and  i think nesting makes things harder to debug and refactor in the future.

If the 2 create project screens are not the same, then I think we can use this, but I think we should split up the createOrJoinProject Screen to have a total of 3 different react components that use this component

cimigree replied:
@ErikSin, I was not able to use one screen for all three iterations of create or name solo project because there is no Project context and no project for when creating a project during on boarding for the first time and it is not possible to conditionally call hooks, as you know.
I know you don't want to nest components at all no matter what, so I had to create two different screens. I could not figure out how to do it without some kind of nesting, so that is why there is now createproject and namesoloproject files now.

ErikSin replied:
> I know you don't want to nest components at all no matter what, so I had to create two different screens. 

I don't actually think this. If you look at my comment above, "If the 2 create project screens are not the same, then I think we can use this, but I think we should split up the createOrJoinProject Screen to have a total of 3 different react components that use this component". 

What i was saying is that **IF** we could avoid nesting, then we should. But since we have to make 3 different screens, we **should** use this component...

ErikSin commented on src/frontend/sharedComponents/Projects/StartNewProjectBase.tsx:
```diff
@@ -0,0 +1,119 @@
+        </View>
+      }>
+      <View style={styles.headerArea}>
+        <TopIcon width={90} height={90} color={DARK_ORANGE} {...iconProps} />
```
if were passing the TopIcon, why doesn't it just deal with props itself. So instead of passing icons and props we just pass an icon that has props set already.

ErikSin replied:
This comment is not relevant anymore as we are getting rid of this componenent

cimigree replied:
Ok no longer relevant.

ErikSin commented on src/frontend/Navigation/Stack/index.tsx:
```diff
@@ -85,7 +98,8 @@ export const RootStackNavigator = () => {
       <RootStack.Navigator {...commonNavigatorProps}>
-        {createProjectOnboardingScreens()}
+        {createProjectOnboardingScreens({intl: formatMessage})}
+        {ErrorBottomSheetGroup}
```
i believe this wont work. I think you need to do `ErrorBottomSheetGroup()` -with the brackets at the end.

But i think to simplify it you can just do:
```ts
const ErrorBottomScreenOption: NativeStackNavigationOptions = {
  presentation: 'transparentModal',
  headerShown: false,
  animation: 'none',
  contentStyle: {backgroundColor: 'transparent'},
};


//in the 2 navigators that use it:
// It also doesn't need to be nested in a `group`. The group just allows for the screens to have shared options, but since there is only one screen we can just apply it directly to the screen
<RootStack.Navigator {...commonNavigatorProps}>
        {...createWhateverScreen()}
        <RootStack.Screen
          name="ErrorBottomSheet"
          options={ErrorBottomScreenOption}
          component={ErrorBottomSheet}
        />
  </RootStack.Navigator>

```

cimigree replied:
You're right the Group isn’t necessary for a single screen. Sorry about that added complication. FWIW the previous version did work because it was a JSX element, not a function and I know that for extra sure because the test I wrote passed. https://github.com/digidem/comapeo-mobile/pull/1334/commits/787691e3c31bd979bcf73d2896698a6f58473ac7

ErikSin reviewed (changes requested):
This is looking good. Sorry, there are some changes, but they are not your fault, they are due to a mistake in the way these were handed off. 

After speaking to aldo, The Start New Project and the Create Project Screen should be exactly the same as when the user is starting a project not in onboarding.

That means we can get rid of your sharedComponents/Project Screens, and just reuse the screens that were already created.

cimigree committed:
Merges with develop

cimigree committed:
Styling tweaks for large type or small screens, gets rid of start new project component

cimigree committed:
Splits create project into two screens to be used by onboarding and regular.

cimigree committed:
Updates tests and messages

cimigree committed:
"Behavior to allow backwards movement from creating a new project in onboarding"

cimigree committed:
Removes error bottom sheet from group.

cimigree committed:
Merge branch 'chore/onboarding-improvements' into feat/project-onboarding

ErikSin commented on src/frontend/screens/Settings/CreateOrJoinProject/CreateOrNameSoloProject/NameSoloProject.tsx:
```diff
@@ -0,0 +1,192 @@
+  );
+}
+
+const styles = StyleSheet.create({
```
If we are not going to use the shared component, then I think we should reuse these styles and share those between the components

cimigree replied:
https://github.com/digidem/comapeo-mobile/pull/1334/commits/c22ce39b1f5c5b7f56fc81851e3029c1235255aa

ErikSin commented on src/frontend/Navigation/Stack/RootStack.test.tsx:
Nice, these test are actually testing the correct behaviour. Good job with these

ErikSin commented on src/frontend/screens/Settings/CreateOrJoinProject/CreateOrNameSoloProject/NameSoloProject.tsx:
```diff
@@ -0,0 +1,192 @@
+
+NameSoloProjectScreen.navTitle = m.navTitle;
+
+type InfoRowProps = {Icon: React.FC<SvgProps>; text: string};
+function InfoRow({Icon, text}: InfoRowProps) {
+  return (
+    <View style={styles.infoRow}>
+      <Icon width={20} height={26} style={styles.infoIcon} />
+      <BodyText variant="smallMeta" style={styles.infoText}>
+        {text}
+      </BodyText>
+    </View>
+  );
+}
```
If you don't want to use the shared screen, then i think this should be reuseble since they are identical

cimigree replied:
https://github.com/digidem/comapeo-mobile/pull/1334/commits/c22ce39b1f5c5b7f56fc81851e3029c1235255aa

ErikSin reviewed (changes requested):
Just to reiterate, Im not against nested components/reusable components, I think we just need to not do it unnecessarily as would have been the case if we could have just used one screen and registered with 3 different nav components.

My suggestion was to use 1 reusable component for the 3 screens... But if we are going to keep it as 2 different screens, I think we should reuse the styles, and reuse some of the smaller pieces as per my comments

Also, can we add screenshots to the PR?

The main blocker though is the create project in the onboarding flow does not have a success screen. I know we are conditionally rendering, and that is why, but I think there is a simple solution.

in RootStackNavigator:

```ts
const [hasInitialized, setHasInitialize] = React.useState(false);

React.useEffect(() => {
    if (!hasInitialized && !storedActiveProjectId && fallbackProjectId) {
      setActiveProjectId(fallbackProjectId)
      setHasInitialized(true)
    } else {
         if(!hasInitialized){
            setHasInitialized(true)
         }
    }

  
  }, [storedActiveProjectId, fallbackProjectId, setActiveProjectId]);


if (!hasInitialized) {
    return <Loading />;
  }

  if (!deviceInfo?.name) {
    return (
        {createOnboardingScreens({intl: formatMessage})}
    );
  }

  // instead of checking for list of project just check for active project id/
  if (!activeProjectId) {
    return (
        {createProjectOnboardingScreens({intl: formatMessage})}
    );
  }

  return (
    createDefaultScreenGroup({intl: formatMessage})
  );

```

Then in the `OnboardingCreateProject` screen DONT set the active project id when the project is created, instead navigate to the success screen and pass the projectID to the success screen. And when the user presses the `Done` on the success screen, we can set the active project id.

cimigree commented:
This is what I said about lack of a Success Screen above in the original PR comments.

> Calls useCreateProject, sets active project on success, and then because there is now a project, the MapScreen and regular app stack populates automatically.
> I have no information in the ticket about which screens I should add next, though there is UX and info from design for the future (I think?). So I was not sure what to do here.

Your PR with the Project Stats opt in is supposed to be the next screen (see Figma), so I didn't think it made a lot of sense to wire up something when your PR would be what is used.

But I can wire up a success screen solution to use in case the Project Stats opt in isn't ready for a while as you say.

ErikSin commented:
> But I can wire up a success screen solution to use in case the Project Stats opt in isn't ready for a while as you say.

Yes, i think it makes sense to wire up the success screen because even with the project stats, it will have to follow a similar architecture...

cimigree committed:
Adds success screen for project created. Some shared styles and shared component.

cimigree commented:
I changed the Root Stack to use the project id. I did not set the project ID for the onboarding when the project is created. I added a new success screen to use temporarily as I did not want to mess with the existing project created success screen (it is a bit complicated already) since the project created success in onboarding will be changing anyway with future work on the project statistics opt in (and it was not at all in scope for this Issue/ PR).

Screenshots:
<img width="250" alt="projectsintroscreen" src="https://github.com/user-attachments/assets/d30405cd-c920-482c-9c82-cc9d326dc6c7" />
<img width="250" alt="startnewproject" src="https://github.com/user-attachments/assets/e6cbf690-d9b9-408e-afa7-82b3a74998d3" />
<img width="250" alt="createproject" src="https://github.com/user-attachments/assets/bc44a549-0655-4488-941f-1bd0b69864cd" />
<img width="250" alt="Success" src="https://github.com/user-attachments/assets/2387927c-01fb-4e86-a494-41bbbaf8f6bc" />
<img  width="250" alt="gotomap" src="https://github.com/user-attachments/assets/ba1f47bf-dfb2-41ea-b1aa-63a75678eeb5" />
<img width="250" alt="joinaproject" src="https://github.com/user-attachments/assets/930dc3d3-9f0f-4182-befe-d7261e388450" />

cimigree committed:
Adds messages. Updates test.

ErikSin commented on src/frontend/Navigation/Stack/index.tsx:
```diff
@@ -38,13 +46,22 @@ export const RootStackNavigator = () => {
+      if (!storedActiveProjectId && fallbackProjectId) {
+        setActiveProjectId(fallbackProjectId);
+      }
+      setHasInitialized(true);
```
You need a loading state. Otherwise there could be a case where there is a `fallbackPorjectId` but no `storedActiveProjectId`, in which case `createProjectOnboardingScreens` will initially render, but then this use effect will be called, and then storedActiveProjectId will be set, causing the `createDefaultScreenGroup` to render.

ErikSin reviewed (approved):
Just need to add a loading state, but othewise looks good. You can merge this into your other PR. That other PR can then be merged into develop, but can you make sure to check the other PR after this is merged so that nothing got lost in the merge

cimigree committed:
Adds a loader for the root stack.
